hooks/shotgun_get_valid_published_file.py

```diff
@@ -0,0 +1,99 @@
+# Copyright (c) 2019 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+Hook with two methods to get a PublishedFile from a single published file
+or a list of published files (legacy TankPublishedFile supported).
+
+This implementation returns the first published file it finds
+matching extensions in the order of the valid_extensions parameter.
+
+If none are found, it raises a TankError.
+"""
+
+import sgtk
+from sgtk import TankError
+from sgtk.util import PublishPathNotDefinedError, PublishPathNotSupported
+
+HookBaseClass = sgtk.get_hook_baseclass()
+
+
+class GetValidPublishedFile(HookBaseClass):
+    """
+    """
+
+    def resolve_single_file(self, entity_type, published_files):
+        """
+        Decide wether or not to return the published file, or raise a TankError.
+        Returns the published file only if its path's extension matches one of
+        the valid_extensions.
+
+        :param str entity_type: PublishedFile or TankPublishedFile.
+        :param list published_files: A list of published files, typically
+               containing only one element.
+        :returns: The published file with the right fields.
+        :raises: TankError, PublishPathNotDefinedError, PublishPathNotSupported
+        """
+        published_file = published_files[0]
+        valid_extensions = self.parent.get_setting("valid_extensions")
+        if not valid_extensions:
+            raise TankError(
+                "Sorry, valid extensions must be provided."
```

Comment:
Since this is a log message and not a UI message, I think being more terse and explicit is better, something like:

```suggestion
                "Missing required value for setting 'valid_extensions'."
```